lib/shoehorn/handler.ex

```diff
@@ -1,21 +1,175 @@
 defmodule Shoehorn.Handler do
-  @callback init() :: any
+  @moduledoc """
+  A behaviour module for implementing handling of failing applications
 
-  @callback application_stopped(app :: atom, reason :: any, state :: any) ::
-              {:noreply, state :: any}
-              | {:stop, reason :: term, state :: any}
+  A Shoehorn.Handler is a module that knows how to respond to specific
+  applications going down. There are two types of failing applications.
+  The first is an application that fails to initialize and the second
+  is an application that stops while it is running.
 
-  defmacro __using__(_) do
+  ## Example
+
+  The Shoehorn.Handler behaviour requires developers to implement two
+  callbacks.
+
+  The `init` callback sets up any state the handler needs.
+
+  Ther `application_started` callback is called when an application
+  starts up.
+
+  The `application_exited` callback processes the incoming failure and
+  replies with the action that `Shoehorn` should take in case of
+  application failure.
+
+          defmodule Example.ShoehornHandler do
+            use Shoehorn.Handler
+
+            def init(_opts) do
+              {:ok, %{restart_counts: 0}}
+            end
+
+            def application_started(app, state) do
+              IO.inspect app
+              {:continue, state}
+            end
+
+            def application_exited(:non_essential_app, _reason,  state) do
+              {:continue, state}
+            end
+
+            def application_exited(:essential_app, _reason, %{restart_counts: restart_counts} = state) when restart_counts < 2 do
+              Application.ensure_all_started(:essential_app)
+              {:continue, %{state | restart_counts: restart_counts + 1}}
+            end
+
+            def applicaton_exited(_, state) do
+              {:halt, state}
+            end
+          end
+
+  We initialize our `Shoehorn.Handler` with a restart count for state
+  by calling `init` with the configuration options from our shoehorn
+  config. This state is stored and passed in from the
+  `Shoehorn.ApplicationController`.
+
+  When we have an application start up we will put a message in the
+  console to notify the developer.
+
+  When we have an non-essential application fail we return `:continue` to
+  inform the system to keep going like nothing happened.
+
+  We restart the essential application of our system two times, and
+  then we tell the system to halt if restarting wasn't fixing the
+  system.
+  """
+
+  @typedoc """
+  The action letting `Shoehorn.ApplicationController` know what to do
+
+  * `:continue` - keep the system going like nothing happened
```

Comment:
`:continue` -> `:ignore`?

Also, `:halt` doesn't need to stop the application first (I actually assume that it doesn't). It just brings the system down, right?

I might have missed it, but is there an advantage to having the app call `:erlang.halt()` vs. returning `{:halt, state}`.

Reply:
True, if you want to halt you could just directly call `:erlanf.halt()` here. Which would mean that `continue` is _any_ response. So we might as well either just expect {:ok, state} or even just state.

Reply:
There might be a benefit to have a seam here in case we find that in the future we needed to so something just prior to calling `:erlang.halt()`

Reply:
Just for consistency I think its good to have the :halt return value.  This is especially true if you want to support multiple handlers in the future

Reply:
I was going to suggest just returning `state` and deleting everything checking the return values to simplify the code. 

Regarding returning `:halt` and having `Shoehorn` call `:erlang.halt()` vs. calling `:erlang.halt()`, I like the latter since it leaves no question to anyone reading the code about what's going to happen next. As for allowing the future option of inserting code between the `:halt` return and calling `:erlang.halt()`, my counter would be that that is project-specific code and you might as well put it in your restart handler. Plus, I'd be concerned that any code in between me requesting a halt and there actually being a halt might hang the system or do something against my intentions.

Reply:
I like the idea of the handler returning events for the main application to execute. The reason is that it keeps the handler as a side effect free piece of software. I was thinking of suggesting bringing back a third return type. `{:foo, fun(), handler_state}`. This return type would make the code that the end user writes simpler to test. It also allows us to later move to a list of handlers without users being required to rewrite how their handlers work. We could sift through the list of handler responses. If one is `:halt` we halt. Otherwise, we run any of the functions that in the return types.

All the side affect code would be fired from one place.